File: Sample-code/WILLDevicesSample-ObjC/WILLDevicesSample-ObjC/ConnectToDeviceViewController.m
//
//  ConnectToDeviceViewController.m
//  WILLDevicesSample-ObjC
//
//  Created by Joss Giffard-Burley on 08/12/2017.
//  Copyright © 2017 Wacom. All rights reserved.
//

#import "ConnectToDeviceViewController.h"
#import "AppDelegate.h"
@import WILLDevicesCore;
@import WILLDevices;

@interface ConnectToDeviceViewController ()

/// The InkDeviceWatcher scans for all ink capture devices and returns device information for connection
@property (nonatomic, strong) InkDeviceWatcher *inkWatcher;

/// The table view that is used to
@property (nonatomic, strong) IBOutlet UITableView *deviceTable;

/// The connect to device button
@property (nonatomic, strong) IBOutlet UIButton *connectButton;

/// The collection of currently discoverable devices
@property (nonatomic, strong) NSMutableArray<InkDeviceInfo *> *discoveredDevices;

/// Info of the current connecting device 
@property (nonatomic, strong) NSObject<InkDevice> * _Nullable connectingInkDevice;


@end

@implementation ConnectToDeviceViewController

//========================================================================================================
// MARK: UIView Methods
//========================================================================================================

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.inkWatcher = [InkDeviceWatcher new];
    self.discoveredDevices = [NSMutableArray<InkDeviceInfo *> new];
    self.inkWatcher.delegate = self;
    self.deviceTable.layer.cornerRadius = 2.5;
    self.deviceTable.tableFooterView = [UIView new];
    
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self.inkWatcher stop];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    self.connectingInkDevice = NULL;
    //Reset any existing data
    [self.inkWatcher reset];
    //Start scanning for devices
    [self.inkWatcher start];
}

//========================================================================================================
// MARK: Button actions
//========================================================================================================

- (IBAction)connectButtonTapped {
    NSIndexPath *selectedPath = self.deviceTable.indexPathForSelectedRow;
    
    if(selectedPath == NULL) {
        return;
    }
    
    NSInteger idx = selectedPath.row;
    
    if(idx >= [self.discoveredDevices count]) {
        return;
    }
    
    @try {
        InkDeviceInfo *connectingDevice = self.discoveredDevices[idx];
        [self.inkWatcher stop];
        [self.discoveredDevices removeAllObjects];
        [self.deviceTable reloadData];
        NSError *err;
        
        self.connectingInkDevice = (NSObject<InkDevice> *)[InkDeviceManager connectToDevice:connectingDevice appID:@"CDLTest" error:&err deviceStatusChangedHandler:^(DeviceStatus oldStatus,  DeviceStatus newStatus) {
            NSString *title = @"Device Status Changed";
            NSString *message;
            NSString *messageID;
            
            switch(newStatus) {
                case DeviceStatusBusy:
                case DeviceStatusNotConnected:
                    return;
                    break;
                case DeviceStatusIdle:
                    message = @"Device connected";
                    messageID = @"connected";
                    self.connectingInkDevice.deviceStatusChanged = nil;
                    [self updateInkDevice];
                    break;
                case DeviceStatusSyncing:
                    message = @"Device syncing";
                    break;
                case DeviceStatusConnecting:
                    message = @"Device connecting";
                    break;
                case DeviceStatusExpectingButtonTapToConfirmConnection:
                    message = @"Tap device button to confirm connection";
                    break;
                case DeviceStatusExpectingButtonTapToReconnect:
                    message = @"Tap device button to reconnect";
                    break;
                case DeviceStatusHoldButtonToEnterUserConfirmationMode:
                    message = @"Hold button to enter user confirmation mode";
                    break;
                case DeviceStatusAcknowledgeConnectionCofirmationTimeout:
                    message = @"Tap device button to acknowledge user timeout";
                    break;
                case DeviceStatusFailedToConnect:
                    message = @"Failed to connect to device. Restarting scan.";
                    [self.inkWatcher start];
                    break;
                case DeviceStatusFailedToPair:
                    message = @"Failed to pair to device. Restarting scan.";
                    [self.inkWatcher start];
                    break;
                case DeviceStatusFailedToAuthorize:
                    message = @"Failed to authorize device. Restarting scan.";
                    [self.inkWatcher start];
                    break;
            }
            if(messageID != NULL) {
                [AppDelegate postNotification:title withBody:message messageID:messageID];
            } else {
                [AppDelegate postNotification:title withBody:message];

            }
        }];

    } @catch(NSException *e) {
        [AppDelegate postNotification:@"Error connecting to device" withBody:e.reason];
        [self.navigationController popToRootViewControllerAnimated:YES];
    }
}
/// Updates the ink device on rootVC
- (void)updateInkDevice {
    if(self.connectingInkDevice != NULL) {
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(3.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            self.rootVC.currentInkDevice = self.connectingInkDevice;
            [self.navigationController popToRootViewControllerAnimated:YES];
        });
    }
}

//========================================================================================================
// MARK: InkDeviceWatcher delegate methods
//========================================================================================================


- (void)deviceAdded:(InkDeviceWatcher * _Nonnull)watcher device:(InkDeviceInfo * _Nonnull)device {
    [self.discoveredDevices addObject:device];
    [self.deviceTable reloadData];
}

- (void)deviceRemoved:(InkDeviceWatcher * _Nonnull)watcher device:(InkDeviceInfo * _Nonnull)device {
    NSPredicate *predicate = [NSPredicate predicateWithBlock:^BOOL(id  _Nullable evaluatedObject, NSDictionary<NSString *,id> * _Nullable bindings) {
        return ![device isEqual:evaluatedObject];
    }];
    [self.discoveredDevices filterUsingPredicate:predicate];
    [self.deviceTable reloadData];
}

//========================================================================================================
// MARK: UITableView delegate methods
//========================================================================================================

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    if([self.discoveredDevices count] > 0) {
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"deviceCell"];
        if(cell == NULL) {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"deviceCell"];
        }
        
        cell.textLabel.text = self.discoveredDevices[indexPath.row].name;
        return(cell);
    } else {
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"blank"];
        if(cell == NULL) {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"blank"];
        }
        return(cell);
    }
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if([self.discoveredDevices count] == 0) {
        return(1);
    } else {
        return([self.discoveredDevices count]);
    }
}

- (nullable NSIndexPath *)tableView:(UITableView *)tableView willSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    if([tableView indexPathForSelectedRow] == indexPath) {
        [tableView deselectRowAtIndexPath:indexPath animated:YES];
        [self.connectButton setEnabled:NO];
        return NULL;
    } else {
        [self.connectButton setEnabled:YES];
        return(indexPath);
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    if([self.discoveredDevices count] > 0) {
        return(44);
    } else {
        return(tableView.frame.size.height);
    }
}

@end

File: Sample-code/WILLDevicesSample/WILLDevicesSample/ConnectToDeviceViewController.swift
//
//  ConnectToDeviceViewController.swift
//  WILLDevicesSample
//
//  Created by Joss Giffard-Burley on 17/07/2017.
//  Copyright © 2017 Wacom. All rights reserved.
//

import UIKit
import WILLDevices

/// Start a new ink watcher and scan for new devices. If the user selects a device in the table view and taps connect, attempt to connect to the selected device.
/// To discover a Smart Pad device during the scan, you will need to hold the button on the device for 6 seconds to allow the device to enter discovery mode
class ConnectToDeviceViewController: UIViewController {
    //========================================================================================================
    // MARK: Properties
    //========================================================================================================

    /// The main VC for the app. This is where we we return the connected ink device
    weak var rootVC: ViewController?
    
    /// The InkDeviceWatcher scans for all ink capture devices and returns device information for connection
    let inkWatcher = InkDeviceWatcher()
    
    /// The table view that is used to
    @IBOutlet var deviceTable: UITableView!
    
    /// The connect to device button
    @IBOutlet var connectButton: UIButton!
    
    /// The collection of currently discoverable devices
    var discoveredDevices = [InkDeviceInfo]()
    
    var connectingInkDevice: InkDevice?
    
    //========================================================================================================
    // MARK: UIView Methods
    //========================================================================================================

    override func viewDidLoad() {
        super.viewDidLoad()
        
        //Set the inkwatcher delegate
        inkWatcher.delegate = self
        deviceTable.layer.cornerRadius = 2.5
        deviceTable.tableFooterView = UIView()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        //Stop the scanner
        inkWatcher.stop()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        //Start scanning for devices
        connectingInkDevice = nil
      
        //Reset any existing
        inkWatcher.reset()
        
        //Start scan
        inkWatcher.start()
    }

    //========================================================================================================
    // MARK: Button actions
    //========================================================================================================

    @IBAction func connectButtonTapped() {
        //Get the selected device info
        guard let idx = deviceTable.indexPathForSelectedRow?.row, idx < discoveredDevices.count else {
            return
        }
        
        let deviceInfo = discoveredDevices[idx]
    
        //Attempt to create a new ink device using the inkdevice factory. The device status change block will
        //be automatically assigned to the underlying InkDevice in the case of successful connection. Possible
        //errors that can be thrown by this call are connection errors and license errors.
        
        do {
            inkWatcher.stop()
            discoveredDevices.removeAll()
            deviceTable.reloadData()
            
            connectingInkDevice = try InkDeviceManager.connectToDevice(deviceInfo, appID: "CDLTestApp", deviceStatusChangedHandler: { [weak self] (oldStatus, newStatus) -> (Void) in
                let title = "Device Status Changed"
                let message :String
                var id :String?
                switch newStatus {
                case .notConnected, .busy:
                    return
                case .idle:
                    message = "Device connected"
                    id = "connected"
                    self?.connectingInkDevice?.deviceStatusChanged = nil
                    self?.updateInkDevice()
                    
                case .syncing:
                    message = "Device syncing"
                case .connecting:
                    message = "Device connecting"
                case .expectingButtonTapToConfirmConnection:
                    message = "Tap device button to confirm connection"
                case .expectingButtonTapToReconnect:
                    message = "Tap device button to reconnect"
                case .holdButtonToEnterUserConfirmationMode:
                    message = "Hold button to enter user confirmation mode"
                case .acknowledgeConnectionCofirmationTimeout:
                    message = "Tap device button to acknowledge user timeout"
                case .failedToConnect:
                    message = "Failed to connect to device. Restarting scan."
                    self?.inkWatcher.start()
                case .failedToPair:
                    message = "Failed to pair to device. Restarting scan."
                    self?.inkWatcher.start()
                case .failedToAuthorize:
                    message = "Failed to authorize device. Restarting scan."
                    self?.inkWatcher.start()
                }
                if id != nil {
                    AppDelegate.postNotification(title, bodyText: message, id:id!)
                } else {
                    AppDelegate.postNotification(title, bodyText: message)
                }
            })
        } catch let e {
            AppDelegate.postNotification("Error connecting device", bodyText: e.localizedDescription)
            navigationController?.popToRootViewController(animated: true)
        }
        
    }
    
    
    /// Updates the ink device on the rootVC
    func updateInkDevice() {
        if connectingInkDevice != nil {
        DispatchQueue.main.asyncAfter(deadline: .now() + 3.0) { [weak self] in
            self?.rootVC?.currentInkDevice = self?.connectingInkDevice
            self?.navigationController?.popToRootViewController(animated: true)
        }
        }
    }

}

//========================================================================================================
// MARK: InkDeviceWatcher delegate methods
//========================================================================================================

extension ConnectToDeviceViewController: InkDeviceWatcherDelegate {
    /// A new `InkDevice` was detected by the ink watcher
    ///
    /// - Parameters:
    ///   - watcher: The watcher that detected the device
    ///   - device: The device info that discovered
    func deviceAdded(_ watcher: InkDeviceWatcher, device: InkDeviceInfo) {
        discoveredDevices.append(device)
        deviceTable.reloadData()
    }

    /// A previously discovered device has been disconnected.
    ///
    /// - Parameters:
    ///   - watcher: The watcher that dected the device
    ///   - device: The device that was discovered
    func deviceRemoved(_ watcher: InkDeviceWatcher, device: InkDeviceInfo) {
        discoveredDevices = discoveredDevices.filter { $0 != device }
        deviceTable.reloadData()
    }
}

//========================================================================================================
// MARK: UITableView delegate methods
//========================================================================================================

extension ConnectToDeviceViewController: UITableViewDelegate, UITableViewDataSource {
    func tableView(_ tableView: UITableView, willSelectRowAt indexPath: IndexPath) -> IndexPath? {
        if tableView.indexPathForSelectedRow == indexPath {
           tableView.deselectRow(at: indexPath, animated: true)
            connectButton.isEnabled = false
            return nil
        } else {
            connectButton.isEnabled = true
            return indexPath
        }
    }
    

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if discoveredDevices.count == 0 { //Create the empty table view
            return 1
        } else {
            return discoveredDevices.count
        }
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if discoveredDevices.count > 0 {
            let cell = tableView.dequeueReusableCell(withIdentifier: "deviceCell", for:indexPath)
            cell.textLabel?.text = discoveredDevices[indexPath.row].name
            return cell
        } else {
           return tableView.dequeueReusableCell(withIdentifier: "blankCell", for:indexPath)
        }
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        if discoveredDevices.count > 0 {
            return 44.0
        } else {
            return tableView.frame.size.height
        }
    }
}

